synthesizer/src/vm/mod.rs

```diff
@@ -107,25 +112,37 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
                         bail!("Transaction id for '{program_id}' is not found in storage.");
                     };
 
-                    // Recursively load the deployment and its imports.
-                    load_deployment_and_imports(process, transaction_store, transaction_id)?
+                    // Add the deployment and its imports found recursively.
+                    deployments.extend_from_slice(&load_deployment_and_imports(
+                        process,
+                        transaction_store,
+                        transaction_id,
+                    )?);
                 }
             }
 
-            // Load the deployment if it does not exist in the process yet.
-            if !process.contains_program(program_id) {
-                process.load_deployment(&deployment)?;
-            }
+            // Once all the imports have been included, add the parent deployment.
+            deployments.push((*program_id, deployment));
 
-            Ok(())
+            Ok(deployments)
         }
 
         // Retrieve the transaction store.
         let transaction_store = store.transaction_store();
         // Load the deployments from the store.
-        for transaction_id in transaction_store.deployment_transaction_ids() {
-            // Load the deployment and its imports.
-            load_deployment_and_imports(&mut process, transaction_store, *transaction_id)?;
+        let deployments = transaction_store
+            .par_deployment_transaction_ids()
+            .map(|transaction_id| {
+                // Load the deployment and its imports.
+                load_deployment_and_imports(&process, transaction_store, *transaction_id)
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        for (program_id, deployment) in deployments.iter().flatten() {
```

Comment:
I think this iterator doesn't need to collect.

This will consume a ton of memory (possibly overflowing memory) if we collect all of the deployments in one go.

Reply:
the reason for the collect is that `rayon` processes those and they couldn't be loaded sequentially otherwise; perhaps we could look into chunking the list of deployment transaction ids in order to limit maximum memory use?

Reply:
updated the PR with a possible approach; it will need to be benchmarked again, so I'll mark the PR as a draft for now